lib/nx_deflate.c

```diff
@@ -745,16 +743,25 @@ static int nx_copy_fifo_out_to_nxstrm_out(nx_streamp s)
 */
 
 /*
-   When small number of bytes are in strm, copy them to fifo instead of
-   using NX to DMA them.  Returns number of bytes copied.
+   When small number of bytes are in strm, copy them to fifo_in instead of
+   using NX to DMA them.
 */
-static inline void small_copy_nxstrm_in_to_fifo_in(nx_streamp s)
+static inline void cache_input(nx_streamp s)
 {
 	uint32_t free_bytes, copy_bytes;
 
-	prt_info("%s:%d avail_in %d used_in %d\n", __FUNCTION__, __LINE__, s->avail_in, s->used_in);
+	/* Avoid underflow if more than half of fifo_in has been used already. */
+	free_bytes = s->cur_in + s->used_in < s->len_in/2 ?
+		s->len_in/2 - s->cur_in - s->used_in : 0;
 
-	free_bytes = s->len_in/2 - s->cur_in - s->used_in;
+	/* Not enough space to cache input. Time to throw away old data, keeping at
+	 * most DEF_HIST_LEN bytes of history. */
+	if(free_bytes < s->avail_in) {
+		memmove(s->fifo_in, s->fifo_in + s->cur_in - DEF_HIST_LEN,
+				DEF_HIST_LEN + s->used_in);
```

Comment:
If we're reaching this point, that means the request had a small number of bytes.
Is it really worth doing this here?
See my next comment.